src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/PublicContract.cs

```diff
@@ -0,0 +1,162 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Utilities;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Helpers used for public API argument validation.
+    /// </summary>
+    internal static class PublicContract
+    {
+        // Guidance on inlining:
+        // Inline implementation of condition checking but don't inline the code that is only executed on failure.
+        // This approach makes the common path efficient (both execution time and code size) 
+        // while keeping the rarely executed code in a separate method.
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void RequireNonNullItems<T>([NotNull] IEnumerable<T>? sequence, string argumentName) where T : class
+        {
+            if (sequence == null)
+            {
+                throw new ArgumentNullException(argumentName);
+            }
+
+            if (sequence.Contains((T)null!))
```

Comment:
Do I dare ask why the (T) cast is necessary here?

Reply:
![image](https://user-images.githubusercontent.com/41759/77581811-9ced3900-6e9b-11ea-80e5-80369d5a7542.png)

Reply:
Oh right I've ran into that too. :frowning: